yaksh/views.py

```diff
@@ -873,34 +855,53 @@ def download_csv(request, questionpaper_id):
 
 
 @login_required
-def grade_user(request, username, questionpaper_id=None):
+def grade_user(request, quiz_id=None, user_id=None, attempt_number=None):
     """Present an interface with which we can easily grade a user's papers
     and update all their marks and also give comments for each paper.
     """
     current_user = request.user
     ci = RequestContext(request)
     if not current_user.is_authenticated() or not is_moderator(current_user):
         raise Http404('You are not allowed to view this page!')
-    data = get_user_data(username, questionpaper_id)
-    if request.method == 'POST':
+    course_details = Course.objects.filter(creator=current_user)
+    context = {"course_details": course_details}
+    if quiz_id is not None:
+        questionpaper_id = QuestionPaper.objects.filter(quiz_id=quiz_id)\
+                                                        .values("id")
+        user_details = AnswerPaper.objects.get_users_for_questionpaper\
+                                            (questionpaper_id)
+        context = {"users": user_details, "quiz_id": quiz_id}
+        if user_id is not None:
+            try:
+                attempts = AnswerPaper.objects.get_user_all_attempts\
+                                                (questionpaper_id, user_id)
+                if attempt_number is None:
+                    attempt_number = attempts[0].attempt_number
+            except IndexError:
+                raise Http404('No attempts for paper')
+            user = User.objects.get(id=user_id)    
+            data = AnswerPaper.objects.get_user_data(user,questionpaper_id,
+                                                     attempt_number
+                                                     )
+
+            context = {'data': data,"quiz_id": quiz_id,  "users": user_details,
+                    "attempts": attempts,"user_id":user_id
+                    }   
+    if request.method == "POST":
         papers = data['papers']
         for paper in papers:
             for question, answers in paper.get_question_answers().iteritems():
                 marks = float(request.POST.get('q%d_marks' % question.id, 0))
-                last_ans = answers[-1]
-                last_ans.marks = marks
-                last_ans.save()
+                answers = answers[-1]
```

Comment:
If possible to do then try else its fine:
It is always better to have a function in models or model manger
For example:
instead of answer[-1] better would be Answers.objects.get_last_answer(question_id)